File: node/mw/xsrf-token.ts
/* @preserve
  (c) 2018 lytical, inc. all rights are reserved.
  lytical(r) is a registered trademark of lytical, inc.
  please refer to your license agreement on the use of this file.
*/

import { RequestHandler } from 'express';
import _uuid = require('uuid');
import HTTP_STATUS_CODES from 'http-status-enum';

export const xsrf_token_cookie_id = 'XSRF-TOKEN';
export const xsrf_token_header_id = 'x-xsrf-token';

export function xsrf_token(): RequestHandler {
  return (rqs, rsp, next) => {
    let cookie = rqs.cookies[xsrf_token_cookie_id];
    if(cookie === undefined) {
      rsp.cookie(xsrf_token_cookie_id, _uuid.v4())
    }
    if(rqs.method !== 'GET' && rqs.method !== 'HEAD') {
      let hdr = rqs.header(xsrf_token_header_id);
      if((hdr || cookie) && hdr !== cookie) {
        rsp.sendStatus(HTTP_STATUS_CODES.BAD_REQUEST);
        return;
      }
    }
    next();
  }
}

File: core/app/xsrf-token.cs
/*
  © 2019 lytical, inc. all rights are reserved.
  lytical® is a registered trademark of lytical, inc.
  please refer to your license agreement on the use of this file.
*/
using Microsoft.AspNetCore.Http;
using System;
using System.Net;

namespace lyt.app
{
  [injectable(typeof(xsrf_token))]
  public interface xsrf_token_i
  {
    public void set_token(HttpResponse response);
    public bool validate_token(HttpContext context, bool is_required = false);
  }

  public class xsrf_token : xsrf_token_i
  {
    public void set_token(HttpResponse rsp) =>
      rsp.Cookies.Append(xsrf_token_cookie_id, Guid.NewGuid().ToString());

    public bool validate_token(HttpContext ctx, bool is_required = false)
    {
      var cookie = ctx.Request.Cookies[xsrf_token_cookie_id];
      if(cookie != null)
      {
        var hdr = ctx.Request.Headers[xsrf_token_header_id];
        if(hdr.Count != 0 && hdr[0] == cookie)
        {
          return true;
        }
      }
      else if(!is_required)
      {
        return true;
      }
      ctx.Response.StatusCode = (int)HttpStatusCode.BadRequest;
      ctx.Response.CompleteAsync();
      return false;
    }

    const string xsrf_token_cookie_id = "XSRF-TOKEN";
    const string xsrf_token_header_id = "x-xsrf-token";
  }
}